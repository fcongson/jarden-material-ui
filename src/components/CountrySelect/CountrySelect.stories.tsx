import * as React from 'react'
import { useState } from 'react'
import { CountrySelect } from './CountrySelect'

export default {
  title: 'Country Select',
  component: CountrySelect
}

export const Single = () => {
  const [value, setValue] = useState('ATA')
  return <CountrySelect value={value} onChange={value => setValue(value)} />
}

export const SingleWithLabel = () => {
  const [value, setValue] = useState()
  return <CountrySelect label={'Country'} value={value} onChange={value => setValue(value)} />
}

export const Multiple = () => {
  const [values, setValues] = useState(['ATA', 'GRL'])
  return <CountrySelect label={'Countries'} value={values} onChange={values => setValues(values)} multiple />
}

export const MultipleWithLabel = () => {
  const [values, setValues] = useState([])
  return <CountrySelect label={'Countries'} value={values} onChange={values => setValues(values)} multiple />
}
