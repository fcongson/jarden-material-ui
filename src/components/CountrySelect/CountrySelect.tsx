import { Checkbox, withStyles } from "@material-ui/core";
import TextField from "@material-ui/core/TextField";
import Check from "@material-ui/icons/Check";
import Close from "@material-ui/icons/Close";
import ExpandMore from "@material-ui/icons/ExpandMore";
import Autocomplete from "@material-ui/lab/Autocomplete";
import * as React from "react";
import { palette } from "../../styles/colors";
import { muiInput, muiSelect, transition } from "../../styles/global.styles";
import { countries } from "./countries";

const mappedCountries: { label: string; value: string }[] = countries.map(
  country => ({
    label: country.name,
    value: country.alphaThree
  })
);

const mapCountry = (country?: string) => {
  const found = countries.filter(({ alphaThree }) => alphaThree === country)[0];
  return {
    label: found?.name ?? "",
    value: found?.alphaThree ?? ""
  };
};

interface CountrySelectProps {
  value?: string;
  onChange?: (value: string) => void | undefined;
}

export const CountrySelect: React.FC<CountrySelectProps> = ({
  value,
  onChange
}) => {
  return (
    <>
      <MaterialUiGlobalCss />
      <Autocomplete
        value={mapCountry(value)}
        autoHighlight
        onChange={(event: any, value: any) => {
          console.log(value);
          if (!!onChange) onChange(value?.value ?? "");
        }}
        options={mappedCountries}
        getOptionLabel={option => option.label}
        getOptionSelected={(option, value) => option.value === value.value}
        renderInput={params => <TextField {...params} fullWidth />}
        popupIcon={<ExpandMore />}
        closeIcon={<Close />}
      />
    </>
  );
};

interface CountrySelectMultipleProps {
  value?: string[];
  onChange?: (value: string[]) => void | undefined;
}

export const CountrySelectMultiple: React.FC<CountrySelectMultipleProps> = ({
  value,
  onChange
}) => {
  const uncheckedIcon = (
    <span className="unchecked">
      <Check
        style={{
          fill: palette["white-base"],
          padding: 2,
          transform: "scale(0)",
          transition: transition
        }}
      />
    </span>
  );
  const checkedIcon = (
    <span className="checked">
      <Check
        style={{
          fill: palette["white-base"],
          padding: 2,
          transform: "scale(1)",
          transition: transition
        }}
      />
    </span>
  );
  const closeIcon = (
    <span
      style={{
        height: 16,
        width: 16,
        borderRadius: 8,
        backgroundColor: palette["white-base"]
      }}
    >
      <Close />
    </span>
  );
  return (
    <>
      <MaterialUiGlobalCss />
      <Autocomplete
        multiple
        disableCloseOnSelect
        value={value?.map(mapCountry)}
        onChange={(event, value) => {
          if (!!onChange)
            onChange(value.map(({ value }: { value: string }) => value));
        }}
        options={mappedCountries}
        getOptionLabel={option => option.label}
        getOptionSelected={(option, value) => option.value === value.value}
        renderOption={(option, { selected }) => (
          <>
            <Checkbox
              icon={uncheckedIcon}
              checkedIcon={checkedIcon}
              style={{}}
              checked={selected}
            />
            {option.label}
          </>
        )}
        renderInput={params => <TextField {...params} fullWidth />}
        popupIcon={<ExpandMore />}
        closeIcon={<Close />}
        ChipProps={{
          deleteIcon: closeIcon
        }}
        classes={{ option: "multiple", inputRoot: "multiple" }}
      />
    </>
  );
};

const MaterialUiGlobalCss = withStyles({
  "@global": {
    '.MuiAutocomplete-inputRoot[class*="MuiInput-root"]': {
      ...muiSelect.size,
      ...muiSelect.spacing,
      ...muiSelect.text.input,
      "& .MuiAutocomplete-input:first-child": {
        ...muiSelect.size,
        ...muiSelect.spacing,
        ...muiSelect.text.input,
        paddingRight: 60
      },
      "&.multiple": {
        paddingRight: 64
      }
    },
    ".MuiInput-underline": {
      "&:not(.Mui-disabled):before, &:hover:not(.Mui-disabled):before":
        muiInput.underline.default,
      "&:after": muiInput.underline.focused
    },
    ".MuiAutocomplete-paper": muiSelect.options,
    ".MuiAutocomplete-listbox": muiSelect.text.option,
    ".MuiAutocomplete-option": {
      "& span.unchecked": {
        ...muiSelect.option.checkbox,
        border: `2px solid ${palette["grey-base"]}`,
        backgroundColor: palette["grey-base"]
      },
      "& span.checked": {
        ...muiSelect.option.checkbox,
        border: `2px solid ${palette["turquioes"]}`,
        backgroundColor: palette["turquioes"]
      },
      '&[data-focus="true"], &[aria-selected="true"]': {
        ...muiSelect.option.selected,
        "& span.unchecked": {
          border: `2px solid ${palette["turquioes"]}`
        },
        "& span.checked": {
          border: `2px solid ${palette["turquioes"]}`
        }
      },
      "&.multiple": {
        ...muiSelect.option.multiple,
        "& .MuiCheckbox-root": {
          padding: "4px 8px 4px 16px"
        },
        "& .MuiCheckbox-colorSecondary.Mui-checked:hover": {
          backgroundColor: "transparent"
        }
      }
    },
    ".MuiIconButton-root": {
      "&:hover": muiSelect.icon.button,
      "& svg": muiSelect.icon.svg
    },
    ".MuiAutocomplete-clearIndicator": muiSelect.icon.button,
    ".MuiAutocomplete-popupIndicator": {
      ...muiSelect.icon.button,
      marginRight: 6
    },
    ".MuiTouchRipple-rippleVisible": muiSelect.icon.animation,
    ".MuiChip-root": {
      ...muiSelect.tag.root,
      "& svg": muiSelect.tag.svg
    },
    ".MuiChip-label": muiSelect.tag.label,
    ".MuiChip-deleteIcon": muiSelect.tag.delete
  }
})(() => null);
