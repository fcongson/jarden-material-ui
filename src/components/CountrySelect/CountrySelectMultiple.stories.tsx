import React, { useState } from 'react'
import { CountrySelectMultiple } from './CountrySelect'

export default {
  title: 'Country Select Multiple',
  component: CountrySelectMultiple
}

export const Default = () => {
  const [values, setValues] = useState(['ATA'])
  return <CountrySelectMultiple value={values} onChange={values => setValues(values)} />
}
