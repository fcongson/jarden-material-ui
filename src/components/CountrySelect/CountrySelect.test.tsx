import { shallow } from 'enzyme'
import * as React from 'react'
import { CountrySelect } from './CountrySelect'

describe('CountrySelect', () => {
  it('Renders', () => {
    const countrySelect = shallow(<CountrySelect />)
    expect(countrySelect).toMatchSnapshot()
  })
})
