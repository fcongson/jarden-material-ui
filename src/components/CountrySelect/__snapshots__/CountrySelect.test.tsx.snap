// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelect Renders 1`] = `ShallowWrapper {}`;
